envoy/adapter/adapter.go

```diff
@@ -0,0 +1,303 @@
+package adapter
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/Nitro/sidecar/catalog"
+	"github.com/Nitro/sidecar/service"
+	api "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	endpoint "github.com/envoyproxy/go-control-plane/envoy/api/v2/endpoint"
+	listener "github.com/envoyproxy/go-control-plane/envoy/api/v2/listener"
+	route "github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+	hcm "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	tcpp "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/tcp_proxy/v2"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/duration"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// ServiceNameSeparator is used to join service name and port. Must not occur in service names.
+	ServiceNameSeparator = ":"
+)
+
+// SvcName formats an Envoy service name from our service name and port
+func SvcName(name string, port int64) string {
+	return fmt.Sprintf("%s%s%d", name, ServiceNameSeparator, port)
+}
+
+// SvcNameSplit an Enovy service name into our service name and port
+func SvcNameSplit(name string) (string, int64, error) {
+	parts := strings.Split(name, ServiceNameSeparator)
+	if len(parts) < 2 {
+		return "", -1, fmt.Errorf("%s", "Unable to split service name and port!")
+	}
+
+	svcName := parts[0]
+	svcPort, err := strconv.ParseInt(parts[1], 10, 64)
+	if err != nil {
+		return "", -1, fmt.Errorf("%s", "Unable to parse port!")
+	}
+
+	return svcName, svcPort, nil
+}
+
+// LookupHost does a vv slow lookup of the DNS host for a service. Totally
+// not optimized for high throughput. You should only do this in development
+// scenarios.
+func LookupHost(hostname string) (string, error) {
+	addrs, err := net.LookupHost(hostname)
+
+	if err != nil {
+		return "", err
+	}
+	return addrs[0], nil
+}
+
+// EnvoyListenersFromState creates a set of Enovy API listener
+// definitions from all the ServicePorts in the Sidecar state.
+func EnvoyListenersFromState(state *catalog.ServicesState, bindIP string) ([]cache.Resource, error) {
+	var listeners []cache.Resource
+
+	state.RLock()
+	defer state.RUnlock()
+
+	svcs := state.ByService()
```

Comment:
I think we can and should use `state.EachService()` which takes a closure to operate on the state items. This save an allocation of a big blob of memory—which will happen pretty often—and it saves an unnecessary sort as well.

A good way to do that would be to build your own hash as you go (it will be _much_ smaller) and then map it into a list at the very end.

```
listenerMap := make(map[string]cache.Resource)

state.EachService(func(hostname *string, id *string, svc *service.Service) {
     // inspect things, put them in the hash as needed, overwriting without checking
})

listeners := make([]cache.Resource, len(listenersMap)
// insert items in a numbered way to avoid using append and the associate re-allocs
```

This is not blocking feedback, but it's a better implementation I think.

Reply:
Hehe, that's what I get for copy/pasting code 😅This is a great suggestion! Now this API should be much faster than the old REST-based one. Thanks for that!